nemo/collections/tts/data/vocoder_dataset.py

```diff
@@ -0,0 +1,181 @@
+
+
+@dataclass
+class DatasetMeta:
+    manifest_path: Path
+    audio_dir: Path
+    sample_weight: float = 1.0
+
+
+@dataclass
+class DatasetSample:
+    manifest_entry: dict
+    audio_dir: Path
+
+
+@experimental
+class VocoderDataset(Dataset):
+    def __init__(
+        self,
+        dataset_meta: Dict[str, DatasetMeta],
+        sample_rate: int,
+        n_segments: Optional[int] = None,
+        weighted_sample_steps: Optional[int] = None,
+        feature_processors: Optional[Dict[str, FeatureProcessor]] = None,
+        min_duration: Optional[float] = None,
+        max_duration: Optional[float] = None,
+        num_audio_retries: int = 5,
+    ):
+        super().__init__()
+
+        self.sample_rate = sample_rate
+        self.n_segments = n_segments
+        self.weighted_sample_steps = weighted_sample_steps
+        self.num_audio_retries = num_audio_retries
+        self.load_precomputed_mel = False
+
+        if feature_processors:
+            logging.info(f"Found feature processors {feature_processors.keys()}")
+            self.feature_processors = feature_processors.values()
+        else:
+            self.feature_processors = []
+
+        self.data_samples = []
+        self.sample_weights = []
+        for dataset_name, dataset in dataset_meta.items():
+            samples, weights = self._process_dataset(
+                dataset_name=dataset_name, dataset=dataset, min_duration=min_duration, max_duration=max_duration,
+            )
+            self.data_samples += samples
+            self.sample_weights += weights
+
+    def get_sampler(self, batch_size: int) -> Optional[torch.utils.data.Sampler]:
+        if not self.weighted_sample_steps:
+            return None
+
+        sampler = get_weighted_sampler(
+            sample_weights=self.sample_weights, batch_size=batch_size, num_steps=self.weighted_sample_steps
+        )
+        return sampler
+
+    def _segment_audio(self, audio_filepath: Path) -> AudioSegment:
+        # Retry file read multiple times as file seeking can produce random IO errors.
+        for _ in range(self.num_audio_retries):
+            try:
+                audio_segment = AudioSegment.segment_from_file(
+                    audio_filepath, target_sr=self.sample_rate, n_segments=self.n_segments,
+                )
+                return audio_segment
+            except Exception:
+                traceback.print_exc()
+
+        raise ValueError(f"Failed to read audio {audio_filepath}")
+
+    def _sample_audio(self, audio_filepath: Path) -> Tuple[torch.Tensor, torch.Tensor]:
+        if not self.n_segments:
+            audio_array, _ = librosa.load(audio_filepath, sr=self.sample_rate)
+        else:
+            audio_segment = self._segment_audio(audio_filepath)
+            audio_array = audio_segment.samples
+        audio = torch.tensor(audio_array)
+        audio_len = torch.tensor(audio.shape[0])
+        return audio, audio_len
+
+    @staticmethod
+    def _process_dataset(
```

Comment:
The name is misleading, this method doesn't process a `Dataset`. Maybe `_preprocess_manifest`?